tools/shared.py

```diff
@@ -1420,6 +1507,63 @@ def open_make_err(i, mode='r'):
     return generated_libs
 
   @staticmethod
+  def make_paths_absolute(f):
+    if f.startswith('-'): # skip flags
+      return f
+    else:
+      return os.path.abspath(f)
+
+  # Runs llvm-nm in parallel for the given list of files.
+  # The results are populated in Building.uninternal_nm_cache
+  # multiprocessing_pool: An existing multiprocessing pool to reuse for the operation, or None
+  # to have the function allocate its own.
+  @staticmethod
+  def parallel_llvm_nm(files):
+    with ToolchainProfiler.profile_block('parallel_llvm_nm'):
+      pool = Building.get_multiprocessing_pool()
+      object_contents = pool.map(g_llvm_nm_uncached, files)
+
+      for i in range(len(files)):
+        Building.uninternal_nm_cache[files[i]] = object_contents[i]
+      return object_contents
+
+  @staticmethod
+  def read_link_inputs(files, just_nm_all_files=False):
+    with ToolchainProfiler.profile_block('read_link_inputs'):
+      # Before performing the link, we need to look at each input file to determine which symbols
+      # each of them provides. Do this in multiple parallel processes.
+      archive_names = [] # .a files passed in to the command line to the link
+      object_names = [] # .o/.bc files passed in to the command line to the link
+      for f in files:
+        absolute_path_f = Building.make_paths_absolute(f)
+
+        # Decide whether this is an archive .a file that we need to extract and enumerate contents of
+        is_archive = Building.is_ar(absolute_path_f)
+        if is_archive and not just_nm_all_files:
+          if not absolute_path_f in Building.ar_contents:
+            archive_names.append(absolute_path_f)
+
+        # Decide whether this is an object file or an .a file that we need to obtain symbol info via llvm-nm.
+        if (not is_archive or just_nm_all_files) and Building.is_bitcode(absolute_path_f):
+          if not absolute_path_f in Building.uninternal_nm_cache:
+            object_names.append(absolute_path_f)
+
+      # Archives contain objects, so process all archives first in parallel to obtain the object files in them.
+      pool = Building.get_multiprocessing_pool()
+      object_names_in_archives = pool.map(extract_archive_contents, archive_names)
+
+      for n in range(len(archive_names)):
+        Building.ar_contents[archive_names[n]] = object_names_in_archives[n]['files']
+
+      for o in object_names_in_archives:
+        for f in o['files']:
+          if not f in Building.uninternal_nm_cache:
+            object_names += [f]
+
+      # Next, extract symbols from all object files (either standalone or inside archives we just extracted)
+      Building.parallel_llvm_nm(object_names)
```

Comment:
return value seems ignored? and doesn't this method return something? or does it have side effects?